File: src/stats.rs
//! Aggregation of per-key observations into namespace -> type statistics.

use serde::Serialize;
use std::cmp::Reverse;
use std::collections::{BinaryHeap, HashMap};

/// Per-type accumulator.
#[derive(Debug, Default, Clone, Serialize)]
pub struct TypeStat {
    pub count: u64,
    /// Total memory in bytes (only populated when `--memory` is enabled).
    pub total_bytes: u64,
    /// Number of keys we actually measured memory for (may be < count if some
    /// keys vanished between SCAN and MEMORY USAGE).
    pub measured: u64,
    /// Breakdown by Redis data type (only when `--types` enabled).
    pub data_types: HashMap<String, u64>,
    /// Keys with no expiry set (only when `--ttl` enabled).
    pub persistent: u64,
    /// Keys with an expiry set.
    pub expiring: u64,
    /// Memory held by persistent (no-TTL) keys — i.e. space that won't be freed
    /// by expiry. Only populated when both memory and TTL are collected.
    pub persistent_bytes: u64,
}

impl TypeStat {
    pub fn avg_bytes(&self) -> u64 {
        self.total_bytes.checked_div(self.measured).unwrap_or(0)
    }
}

/// One observation gathered for a single key.
#[derive(Debug, Default)]
pub struct KeyObservation {
    pub bytes: Option<u64>,
    pub data_type: Option<String>,
    pub ttl: Option<i64>, // -1 = no expiry, -2 = missing, >=0 seconds
}

/// A single key retained in the biggest-keys leaderboard.
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct BigKey {
    pub key: String,
    pub bytes: u64,
    pub data_type: Option<String>,
    pub ttl: Option<i64>,
}

#[derive(Debug, Default, Clone, Serialize)]
pub struct Stats {
    /// namespace -> (type -> stat)
    pub namespaces: HashMap<String, HashMap<String, TypeStat>>,
    pub total_keys: u64,

    /// Bounded leaderboard of the largest keys by memory. Kept as a min-heap
    /// (smallest-on-top) capped at `biggest_cap` so memory stays constant no
    /// matter how huge the DB is. Empty unless memory measurement is enabled.
    #[serde(skip)]
    biggest_heap: BinaryHeap<Reverse<HeapEntry>>,
    #[serde(skip)]
    biggest_cap: usize,
}

/// Heap entry ordered purely by byte size (then key for deterministic ties).
#[derive(Debug, Clone, PartialEq, Eq)]
struct HeapEntry {
    bytes: u64,
    key: String,
    data_type: Option<String>,
    ttl: Option<i64>,
}

impl PartialOrd for HeapEntry {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}
impl Ord for HeapEntry {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.bytes
            .cmp(&other.bytes)
            .then_with(|| self.key.cmp(&other.key))
    }
}

impl Stats {
    /// Create a Stats that retains the `cap` largest keys (0 disables tracking).
    pub fn with_biggest_cap(cap: usize) -> Self {
        Stats {
            biggest_cap: cap,
            ..Default::default()
        }
    }

    pub fn record(&mut self, key: &str, namespace: &str, key_type: &str, obs: &KeyObservation) {
        self.total_keys += 1;
        {
            let ns = self.namespaces.entry(namespace.to_string()).or_default();
            let stat = ns.entry(key_type.to_string()).or_default();
            stat.count += 1;

            if let Some(bytes) = obs.bytes {
                stat.total_bytes += bytes;
                stat.measured += 1;
            }
            if let Some(dt) = &obs.data_type {
                *stat.data_types.entry(dt.clone()).or_insert(0) += 1;
            }
            if let Some(ttl) = obs.ttl {
                if ttl == -1 {
                    stat.persistent += 1;
                    // Track the memory of no-TTL keys: space expiry won't free.
                    if let Some(bytes) = obs.bytes {
                        stat.persistent_bytes += bytes;
                    }
                } else if ttl >= 0 {
                    stat.expiring += 1;
                }
                // ttl == -2 means the key disappeared; don't count it.
            }
        }

        // Done with the per-type borrow; now offer to the global leaderboard.
        if let Some(bytes) = obs.bytes {
            self.record_big_key(key, bytes, &obs.data_type, obs.ttl);
        }
    }

    /// Offer a measured key to the bounded biggest-keys leaderboard.
    fn record_big_key(
        &mut self,
        key: &str,
        bytes: u64,
        data_type: &Option<String>,
        ttl: Option<i64>,
    ) {
        if self.biggest_cap == 0 {
            return;
        }
        // Skip work once full if this key can't beat the current smallest.
        if self.biggest_heap.len() >= self.biggest_cap {
            if let Some(Reverse(smallest)) = self.biggest_heap.peek() {
                if bytes <= smallest.bytes {
                    return;
                }
            }
            self.biggest_heap.pop();
        }
        self.biggest_heap.push(Reverse(HeapEntry {
            bytes,
            key: key.to_string(),
            data_type: data_type.clone(),
            ttl,
        }));
    }

    /// The retained biggest keys, sorted largest-first.
    pub fn biggest(&self) -> Vec<BigKey> {
        let mut v: Vec<BigKey> = self
            .biggest_heap
            .iter()
            .map(|Reverse(e)| BigKey {
                key: e.key.clone(),
                bytes: e.bytes,
                data_type: e.data_type.clone(),
                ttl: e.ttl,
            })
            .collect();
        v.sort_by(|a, b| b.bytes.cmp(&a.bytes).then_with(|| a.key.cmp(&b.key)));
        v
    }

    /// Total key count for a namespace.
    pub fn namespace_count(&self, ns: &str) -> u64 {
        self.namespaces
            .get(ns)
            .map(|m| m.values().map(|s| s.count).sum())
            .unwrap_or(0)
    }

    /// Total memory for a namespace.
    pub fn namespace_bytes(&self, ns: &str) -> u64 {
        self.namespaces
            .get(ns)
            .map(|m| m.values().map(|s| s.total_bytes).sum())
            .unwrap_or(0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn obs(bytes: u64, ttl: i64) -> KeyObservation {
        KeyObservation {
            bytes: Some(bytes),
            data_type: Some("string".into()),
            ttl: Some(ttl),
        }
    }

    #[test]
    fn records_and_aggregates() {
        let mut s = Stats::default();
        s.record("ns1:a", "ns1", "product", &obs(100, -1));
        s.record("ns1:b", "ns1", "product", &obs(300, 60));
        assert_eq!(s.total_keys, 2);
        let stat = &s.namespaces["ns1"]["product"];
        assert_eq!(stat.count, 2);
        assert_eq!(stat.total_bytes, 400);
        assert_eq!(stat.avg_bytes(), 200);
        assert_eq!(stat.persistent, 1);
        assert_eq!(stat.expiring, 1);
        // Only the no-TTL key (100 bytes) counts toward unevictable memory.
        assert_eq!(stat.persistent_bytes, 100);
        assert_eq!(stat.data_types["string"], 2);
        assert_eq!(s.namespace_count("ns1"), 2);
    }

    #[test]
    fn biggest_keys_are_bounded_and_sorted() {
        let mut s = Stats::with_biggest_cap(3);
        for (i, bytes) in [50u64, 500, 10, 999, 250, 700].iter().enumerate() {
            s.record(&format!("k{i}"), "ns", "t", &obs(*bytes, -1));
        }
        let big = s.biggest();
        // Only the cap is retained, largest first.
        assert_eq!(big.len(), 3);
        assert_eq!(big[0].bytes, 999);
        assert_eq!(big[1].bytes, 700);
        assert_eq!(big[2].bytes, 500);
    }

    #[test]
    fn biggest_disabled_when_cap_zero() {
        let mut s = Stats::default(); // cap 0
        s.record("k", "ns", "t", &obs(123, -1));
        assert!(s.biggest().is_empty());
    }
}

File: internal/stats.go
package internal

import (
	"container/heap"
	"sort"
)

// TypeStat holds per-type accumulator.
type TypeStat struct {
	Count           int64            `json:"count"`
	TotalBytes      int64            `json:"total_bytes"`
	Measured        int64            `json:"measured"`
	DataTypes       map[string]int64 `json:"data_types,omitempty"`
	Persistent      int64            `json:"persistent"`
	Expiring        int64            `json:"expiring"`
	PersistentBytes int64            `json:"persistent_bytes"`
}

// AvgBytes returns the average bytes per key.
func (t *TypeStat) AvgBytes() int64 {
	if t.Measured == 0 {
		return 0
	}
	return t.TotalBytes / t.Measured
}

// KeyObservation is one observation gathered for a single key.
type KeyObservation struct {
	Bytes    *int64
	DataType *string
	TTL      *int64 // -1 = no expiry, -2 = missing, >=0 seconds
}

// BigKey is a single key retained in the biggest-keys leaderboard.
type BigKey struct {
	Key      string  `json:"key"`
	Bytes    int64   `json:"bytes"`
	DataType *string `json:"data_type,omitempty"`
	TTL      *int64  `json:"ttl,omitempty"`
}

// heapEntry is used for the min-heap of biggest keys.
type heapEntry struct {
	bytes    int64
	key      string
	dataType *string
	ttl      *int64
}

// bigKeyHeap implements heap.Interface (min-heap).
type bigKeyHeap []heapEntry

func (h bigKeyHeap) Len() int           { return len(h) }
func (h bigKeyHeap) Less(i, j int) bool { return h[i].bytes < h[j].bytes }
func (h bigKeyHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *bigKeyHeap) Push(x any)        { *h = append(*h, x.(heapEntry)) }
func (h *bigKeyHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// Stats aggregates per-key observations into namespace -> type statistics.
type Stats struct {
	Namespaces  map[string]map[string]*TypeStat `json:"namespaces"`
	TotalKeys   int64                           `json:"total_keys"`
	biggestHeap bigKeyHeap
	biggestCap  int
}

// NewStats creates a Stats that retains the `cap` largest keys (0 disables tracking).
func NewStats(cap int) *Stats {
	return &Stats{
		Namespaces: make(map[string]map[string]*TypeStat),
		biggestCap: cap,
	}
}

// Record adds a key observation to the stats.
func (s *Stats) Record(key, namespace, keyType string, obs *KeyObservation) {
	s.TotalKeys++

	if s.Namespaces[namespace] == nil {
		s.Namespaces[namespace] = make(map[string]*TypeStat)
	}
	ns := s.Namespaces[namespace]

	if ns[keyType] == nil {
		ns[keyType] = &TypeStat{DataTypes: make(map[string]int64)}
	}
	stat := ns[keyType]
	stat.Count++

	if obs.Bytes != nil {
		b := *obs.Bytes
		stat.TotalBytes += b
		stat.Measured++
	}
	if obs.DataType != nil {
		stat.DataTypes[*obs.DataType]++
	}
	if obs.TTL != nil {
		ttl := *obs.TTL
		if ttl == -1 {
			stat.Persistent++
			if obs.Bytes != nil {
				stat.PersistentBytes += *obs.Bytes
			}
		} else if ttl >= 0 {
			stat.Expiring++
		}
	}

	// Offer to the leaderboard
	if obs.Bytes != nil {
		s.recordBigKey(key, *obs.Bytes, obs.DataType, obs.TTL)
	}
}

func (s *Stats) recordBigKey(key string, bytes int64, dataType *string, ttl *int64) {
	if s.biggestCap == 0 {
		return
	}
	// Skip if full and key can't beat the current smallest.
	if s.biggestHeap.Len() >= s.biggestCap {
		if s.biggestHeap[0].bytes >= bytes {
			return
		}
		heap.Pop(&s.biggestHeap)
	}
	dt := dataType
	var t *int64
	if ttl != nil {
		v := *ttl
		t = &v
	}
	heap.Push(&s.biggestHeap, heapEntry{
		bytes:    bytes,
		key:      key,
		dataType: dt,
		ttl:      t,
	})
}

// Biggest returns the retained biggest keys, sorted largest-first.
func (s *Stats) Biggest() []BigKey {
	result := make([]BigKey, 0, s.biggestHeap.Len())
	for _, e := range s.biggestHeap {
		result = append(result, BigKey{
			Key:      e.key,
			Bytes:    e.bytes,
			DataType: e.dataType,
			TTL:      e.ttl,
		})
	}
	sort.Slice(result, func(i, j int) bool {
		if result[i].Bytes != result[j].Bytes {
			return result[i].Bytes > result[j].Bytes
		}
		return result[i].Key < result[j].Key
	})
	return result
}

// NamespaceCount returns the total key count for a namespace.
func (s *Stats) NamespaceCount(ns string) int64 {
	m, ok := s.Namespaces[ns]
	if !ok {
		return 0
	}
	var total int64
	for _, stat := range m {
		total += stat.Count
	}
	return total
}

// NamespaceBytes returns the total memory for a namespace.
func (s *Stats) NamespaceBytes(ns string) int64 {
	m, ok := s.Namespaces[ns]
	if !ok {
		return 0
	}
	var total int64
	for _, stat := range m {
		total += stat.TotalBytes
	}
	return total
}
